Solution 3 days, Advent of Code 2023

SoulViolin opened the pull request.

SoulViolin committed:
Solution 3 days, Advent of Code 2023

codesnik commented on Quest-3/Part-1/main.py:
```diff
@@ -0,0 +1,69 @@
+    numbers = {}
+
+    def exploreNumber(row, col):
+        if row < 0 or col < 0 or row >= len(matrix) or col >= len(matrix[0]) or (row, col) in visited or matrix[row][col] == '.' or matrix[row][col] in symbols:
```
which one is wrong?

SoulViolin replied:
or matrix[row][col] in symbols

codesnik replied:
yeah, that one is weird. You have no idea what symbols could be there. No need to scan for them beforehand. You can check for not being in "." and digits.

codesnik commented on Quest-3/Part-1/main.py:
```diff
@@ -0,0 +1,69 @@
+def main(fileName):
+    lines = readFile(fileName)
+
+    schemEngine = [list(line.replace("\n", "")) for line in lines]
```
rstrip("\n") is better

codesnik commented on Quest-3/Part-1/main.py:
```diff
@@ -0,0 +1,69 @@
+
+        visited.add((row, col))
+        number = [(row, col)]
+        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
```
wtf. Numbers are not supposed to grow in all four directions.

codesnik commented on Quest-3/Part-1/main.py:
```diff
@@ -0,0 +1,69 @@
+
+if __name__ == '__main__':
+    fileName = "input.txt"
+    main(fileName)
```
btw, do you see that (-) sign? this is beacause your last line is not ending in "\n", and unix text file conventions usually require it.
In vscode you can take care to have one more empty line at the end. or you can turn on that setting: 
![image](https://github.com/SoulViolin/Advent-of-Code-2023/assets/4512/698ca6a5-4892-4a19-b265-7317ae414c1e)

codesnik commented on Quest-3/Part-1/main.py:
```diff
@@ -0,0 +1,69 @@
+    lines = readFile(fileName)
+
+    schemEngine = [list(line.replace("\n", "")) for line in lines]
+    symbols = set(char for line in schemEngine for char in line if char != '.' and not char.isdigit())
```
this is something really weird. I can't parse it. But thankfully it's probably not needed. You could have a set of of coords of symbols, not symbols themselves.

codesnik replied:
reminding you of that. you don't need `symbols` at all. Just imagine your field is very, very big. You just had an additional scan just to count them. This is wasteful.

codesnik commented on Quest-3/Part-1/main.py:
```diff
@@ -0,0 +1,69 @@
+
+    return numbers
+
+def findSymbolsAroundNumbers(matrix, numbers, symbols):
```
too complex. You know the length of each number. Numbers are written horizontally, always. So you can actually compute all the coords of symbols around the number, instead of scanning around each digit.

SoulViolin committed:
The number search function has been rewritten

SoulViolin committed:
Remove unnecessary code

SoulViolin committed:
Remove unnecessary code

SoulViolin committed:
The second part of the task has been completed.

Code changes:
1. The amount is calculated from "gearRatios", not from "trueNumbers";
2. Logic has been added to the findSymbolsAroundNumbers function to save the coordinates of the symbol (58, 68, 74 line).

New code:
The findGearRatios function has been written

SoulViolin committed:
The second part of the task has been completed.

Code changes:
1. The amount is calculated from "gearRatios", not from "trueNumbers";
2. Logic has been added to the findSymbolsAroundNumbers function to save the coordinates of the symbol (58, 68, 74 line).

New code:
The findGearRatios function has been written

codesnik commented on Quest-3/Part-1/main.py:
```diff
@@ -0,0 +1,78 @@
+            else:
+                break
+        # Explore vertically
+        for r in range(row + 1, rows):
```
why vertically? numbers are not vertical in this task.

SoulViolin committed:
Remove unnecessary code

codesnik commented on Quest-3/Part-1/main.py:
```diff
@@ -0,0 +1,71 @@
+
+    for row in range(rows):
+        for col in range(cols):
+            if matrix[row][col].isdigit() and (row, col) not in visited:
```
you check for this in `explore` too. maybe your explore function could do more? and return None if there's no number.

codesnik commented on Quest-3/Part-1/main.py:
```diff
@@ -0,0 +1,71 @@
+                if (newRow, newCol) in coordinates:
+                    continue  # Skip a cell if it is part of the same number
+                if 0 <= newRow < len(matrix) and 0 <= newCol < len(matrix[0]) and matrix[newRow][newCol] in symbols:
+                    foundSymbol = True
```
instead of a flag `foundSymbol` and multiple breaks, you can extract this loop in a function and return True from inside of the loop. This is usually much better. Flags of any sort suck.

SoulViolin committed:
Add doctest

SoulViolin committed:
Edit gitignor

SoulViolin committed:
Change explore

SoulViolin committed:
Changes in the calculation logic

SoulViolin committed:
Total rewrite

SoulViolin committed:
Fixed bug

SoulViolin committed:
Remove unnecessary code

SoulViolin committed:
Add new function overlap

SoulViolin committed:
Change logic

codesnik reviewed (approved):
Ok, Part-2 is now much better.

SoulViolin committed:
Merge branch 'main' into Quest-3
